src/util/codec/mysql/json.rs

```diff
@@ -0,0 +1,515 @@
+            let key_len = try!(key_entries_data.decode_u16_le());
+            let key_data = &data[key_offset..(key_offset + key_len as usize)];
+            let key = String::from(str::from_utf8(key_data).unwrap());
+            let value =
+                try!(value_entries_data.decode_json_item(&data[key_offset..], key_real_offset));
+            obj.insert(key, value);
+            key_offset += key_len as usize;
+        }
+        Ok(Json::Object(obj))
+    }
+
+    fn decode_json_array(&mut self) -> Result<Json> {
+        // count size value_entries values
+        let element_count = try!(self.decode_u32_le()) as usize;
+        let total_size = try!(self.decode_u32_le());
+        // already removed count and size
+        let left_size = total_size as usize - LENGTH_U32 - LENGTH_U32;
+        let mut data = vec![0;left_size];
+        try!(self.read_exact(&mut data));
+        let value_entries_len = LENGTH_VALUE_ENTRY * element_count;
+        let mut value_entries_data = &data[0..value_entries_len];
+        let values_data = &data[value_entries_len..];
+        let mut array_data = Vec::with_capacity(element_count);
+        let data_start_offset = (LENGTH_U32 + LENGTH_U32 + value_entries_len) as u32;
+        for _ in 0..element_count {
+            let value = try!(value_entries_data.decode_json_item(values_data, data_start_offset));
+            array_data.push(value);
+        }
+        Ok(Json::Array(array_data))
+    }
+
+    fn decode_json_str(&mut self) -> Result<Json> {
+        let length = try!(self.decode_var_u64());
+        let mut encode_value = vec![0;length as usize];
+        try!(self.read_exact(&mut encode_value));
+        let value = try!(String::from_utf8(encode_value));
+        Ok(Json::String(value))
+    }
+
+    fn decode_json_literal(&mut self) -> Result<Json> {
+        let l = try!(self.read_u8());
+        Ok(Json::Literal(l))
+    }
+
+    fn decode_json_double(&mut self) -> Result<Json> {
+        let value = try!(self.decode_f64_le());
+        Ok(Json::Double(value))
+    }
+
+    fn decode_json_i64(&mut self) -> Result<Json> {
+        let value = try!(self.decode_i64_le());
+        Ok(Json::I64(value))
+    }
+
+    fn decode_json_item(&mut self, values_data: &[u8], data_start_position: u32) -> Result<Json> {
+        let mut entry = vec![0;LENGTH_VALUE_ENTRY];
+        try!(self.read_exact(&mut entry));
+        let mut entry = entry.as_slice();
+        let code = try!(entry.read_u8());
+        match code {
+            TYPE_CODE_LITERAL => entry.decode_json_body(code),
+            _ => {
+                let real_offset = entry.decode_u32_le().unwrap();
+                let offset_in_values = real_offset - data_start_position;
+                let mut value = &values_data[offset_in_values as usize..];
+                value.decode_json_body(code)
+            }
+        }
+    }
+}
+
+impl<T: Read> JsonDecoder for T {}
+
+fn normalize(data: Value) -> Json {
```

Comment:
It seems may make the code structure more complicated since we should not only implement [Deserialize](https://serde.rs/impl-deserialize.html), but also need to implement  [Visiter](https://docs.serde.rs/serde/de/trait.Visitor.html) @BusyJay ?